src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -86,61 +89,101 @@ public void Update(IContainSagaData saga)
 
         void ValidateUniqueProperties(IContainSagaData saga)
         {
-            var sagaMetaData = sagaModel.FindByEntityName(saga.GetType().FullName);
-
-            if (!sagaMetaData.CorrelationProperties.Any()) return;
+            var sagaType = saga.GetType();
+            var sagaMetaData = sagaModel.FindByEntityName(sagaType.FullName);
 
-            var sagasFromSameType = from s in data
-                                    where
-                                        (s.Value.SagaEntity.GetType() == saga.GetType() && (s.Key != saga.Id))
-                                    select s.Value;
+            if (sagaMetaData.CorrelationProperties.Count == 0) return;
 
-            foreach (var storedSaga in sagasFromSameType)
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var storedSaga in data)
             {
+                if (storedSaga.Value.SagaData.GetType() != sagaType || (storedSaga.Key == saga.Id)) continue;
```

Comment:
It was used in other parts in this file as well that's why I kept it consistent.